### PR TITLE
feat: 🎸 new no-wait-for-snapshot rule

### PR DESCRIPTION
**Pull request by Gpx**
✅ Closes: #214

The rule is checking for all async utils not only `wait`. If that's not the preferred behavior I can change it back. It's my first contribution to this project so any feedback is welcome 😄

### CODE REVIEW

File: docs/rules/no-wait-snapshot.test.md
Comment (Belco90) [inline docs/rules/no-wait-snapshot.test.md:3]: Can you make reference to `waitFor` rather than `wait`, and add a note at the end of Rule Details to clarify all util methods this rule reports?

**Review by Belco90 — CHANGES_REQUESTED** [review]
Hey Giorgio, thanks for your first PR! Looking good, I just requested some rewording on the rule doc and should be fine.

We are usually asking to check if corresponding methods are imported from RTL, but I think it's fine with this one until we release v4.

File: docs/rules/no-wait-snapshot.test.md
Comment (Gpx) [reply docs/rules/no-wait-snapshot.test.md:3]: Sure thing

Comment (Gpx) [reply docs/rules/no-wait-snapshot.test.md:3]: Do you want to change the name of the rule too? Something like `no-waitfor-snapshot`?

Comment (Belco90) [reply docs/rules/no-wait-snapshot.test.md:3]: That's a good idea! Let's go for `no-wait-for-snapshot` for consistency with `no-wait-for-empty-callback`

File: lib/node-utils.ts
Comment (Gpx) [inline lib/node-utils.ts:87]: I've added this to avoid errors in case `node.parent` is `null`

File: docs/rules/no-wait-for-snapshot.test.md
Comment (gndelia) [inline docs/rules/no-wait-for-snapshot.test.md:10]: minor spelling suggestions: 
```suggestion
The problem then is the amount of loop ran until the condition is met will vary between different computers (or CI machines). This leads to tests that will regenerate a lot of snapshots until the condition is matched when devs run those tests locally updating the snapshots; e.g devs cannot run `jest -u` locally or it'll generate a lot of invalid snapshots who'll fail during CI.
```

**Comment by gndelia** [comment]
Thanks for your PR !! 🚀 ❤️ 

@Belco90  one doubt I had: should we target new rules directly to v4, in order to simplify conflict resolution once we want to go out with that version? Especially for the changes we're considering to add in all rules

File: lib/rules/no-wait-for-snapshot.ts
Comment (gndelia) [inline lib/rules/no-wait-for-snapshot.ts:23]: I think it'd be better to here use the async utility message that caused the error, rather than the hardcoded `waitFor`.

Otherwise, we'd face scenarios like the following:

```ts
await wait(() => {
    expect(container).toMatchSnapshot();
  });
// output error message: 'Ensures no snapshot is generated inside of a `waitFor` call'
// expected output error message: 'Ensures no snapshot is generated inside of a `wait` call'
```

File: lib/rules/no-wait-for-snapshot.ts
Comment (gndelia) [inline lib/rules/no-wait-for-snapshot.ts:80]: There's a type guard utility function [here](https://github.com/testing-library/eslint-plugin-testing-library/blob/636273a23d9bda48a9d2945abf51df9d1ee93c59/lib/node-utils.ts#L14) that will allow you to drop the  cast

```suggestion
          if (isMemberExpression(usage.node)) {
           // object is automatically a TSESTree.Identifier
            const object = usage.node.object;
```

File: lib/rules/no-wait-for-snapshot.ts
Comment (gndelia) [inline lib/rules/no-wait-for-snapshot.ts:96]: I'd have hoped the linter configuration enforced `!==` over `!=`

I tend to avoid `!=` completely but I'm not sure here 😅 

Thoughts @Belco90 ?

Comment (Gpx) [reply lib/rules/no-wait-for-snapshot.ts:97]: I think the linter allows for `!=` because I'm checking against `null`. If you use `!=` it's implicitly checking for `undefined` too

File: lib/rules/no-wait-for-snapshot.ts
Comment (Gpx) [reply lib/rules/no-wait-for-snapshot.ts:23]: Good call

**Comment by Gpx** [comment]
The rule is now checking for `toMatchInlineSnapshot` too and in the error message it returns which async method was used

**Comment by Belco90** [comment]
> @Belco90 one doubt I had: should we target new rules directly to v4, in order to simplify conflict resolution once we want to go out with that version? Especially for the changes we're considering to add in all rules

It would simplify things on v4, but I don't want to block nice additions just to wait for v4. It will take more work on v4 but that's fine.

File: lib/rules/no-wait-for-snapshot.ts
Comment (Belco90) [reply lib/rules/no-wait-for-snapshot.ts:100]: I think in this case is fine to compare against nil, and I understand that's the actual usage right?

**Review by Belco90 — APPROVED** [review]
Awesome! Nice improvements. I'll wait for @gndelia approval for merging it.

**Comment by Belco90** [comment]
@all-contributors please add @Gpx for code, test and docs.

**Comment by arvigeus** [comment]
The docs file is wrong: `docs/rules/no-wait-for-snapshot.test.md`

I don't think it should have `test` suffix. The The README also points to `docs/rules/no-wait-for-snapshot.md`

**Comment by Belco90** [comment]
@arvigeus fixed in #225 !

**Pull request merged**
